domain/secret/state/state.go

```diff
@@ -560,3 +599,183 @@ WHERE  rev.secret_id = $secretRevision.secret_id AND rev.revision = $secretRevis
+AND   suc.unit_uuid = $secretUnitConsumer.unit_uuid
+`
+
+	queryStmt, err := st.Prepare(query, secretUnitConsumer{})
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	selectUnitUUID := `SELECT &M.uuid FROM unit WHERE unit_id=$M.unit_id`
+	selectUnitUUIDStmt, err := st.Prepare(selectUnitUUID, sqlair.M{})
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	selectLatestRevision := `
+SELECT MAX(revision) AS &M.latest_revision
+FROM secret_revision rev
+WHERE rev.secret_id = $M.secret_id`
+	selectLatestRevisionStmt, err := st.Prepare(selectLatestRevision, sqlair.M{})
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	var (
+		dbSecretConsumers secretUnitConsumers
+		latestRevision    int
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		if err := st.checkExistsIfLocal(ctx, tx, uri); err != nil {
+			return errors.Trace(err)
+		}
+
+		result := sqlair.M{}
+		err = tx.Query(ctx, selectUnitUUIDStmt, sqlair.M{"unit_id": unitName}).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return uniterrors.NotFound
+			} else {
+				return errors.Annotatef(err, "looking up unit UUID for %q", unitName)
+			}
+		}
+		consumer.UnitUUID = result["uuid"].(string)
+		err = tx.Query(ctx, queryStmt, consumer).GetAll(&dbSecretConsumers)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotate(err, "querying secret consumers")
+		}
+
+		// TODO(secrets) - we need something different for cross model secrets
+		err = tx.Query(ctx, selectLatestRevisionStmt, sqlair.M{"secret_id": uri.ID}).Get(&result)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return secreterrors.SecretNotFound
+			} else {
+				return errors.Annotatef(err, "looking up latest revision for %q", uri.ID)
+			}
+		}
+		rev, _ := result["latest_revision"].(int64)
+		latestRevision = int(rev)
+
+		return nil
+	})
+	if err != nil {
+		return nil, 0, errors.Trace(domain.CoerceError(err))
+	}
+	if len(dbSecretConsumers) == 0 {
+		return nil, latestRevision, fmt.Errorf("secret consumer for %q and unit %q%w", uri.ID, unitName, secreterrors.SecretConsumerNotFound)
+	}
+	consumers, err := dbSecretConsumers.toSecretConsumers()
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+	return consumers[0], latestRevision, nil
+}
+
+// SaveSecretConsumer saves the consumer metadata for the given secret and unit.
+// If the unit does not exist, an error satisfying [uniterrors.NotFound] is returned.
+// If the secret does not exist, an error satisfying [secreterrors.SecretNotFound] is returned.
+func (st State) SaveSecretConsumer(ctx context.Context, uri *coresecrets.URI, unitName string, md *coresecrets.SecretConsumerMetadata) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertQuery := `
+INSERT INTO secret_unit_consumer (*)
+VALUES ($secretUnitConsumer.*)
+ON CONFLICT(secret_id, unit_uuid) DO UPDATE SET
+    label=excluded.label,
```

Comment:
Do we want to reset label if the provided is ""?

Reply:
The label is set by the caller.